File: tests/tamm/Test_LabeledTensor.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest/doctest.h"
#include <tamm/tamm.hpp>
#include <iostream>

using namespace tamm;


TEST_CASE("Zero-dimensional tensor") {
  bool failed = false;
  //IndexSpace is {range(10)};
  //TiledIndexSpace tis{is};
  Tensor<double> T1{};
  try {
    auto lt = T1();
  } catch (...) {
    failed = true;
  }
  REQUIRE(!failed);
}

TEST_CASE("Zero-dimensional tensor wrong str count") {
  bool failed = false;
  //IndexSpace is {range(10)};
  //TiledIndexSpace tis{is};
  Tensor<double> T1{};
  try {
    LabeledTensor<double> lt{T1("a")};
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
}

TEST_CASE("Zero-dimensional tensor wrong label count") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{};
  try {
    LabeledTensor<double> lt{T1(tis)};
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
}

TEST_CASE("One-dimensional tensor") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{tis};
  try {
    auto lt = T1();
  } catch (...) {
    failed = true;
  }
  REQUIRE(!failed);
}

TEST_CASE("One-dimensional tensor with correct string count") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{tis};
  try {
    auto lt = T1("i");
  } catch (...) {
    failed = true;
  }
  REQUIRE(!failed);
}

TEST_CASE("One-dimensional tensor with correct label count") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{tis};
  try {
    auto lt = T1(tis.label());
  } catch (...) {
    failed = true;
  }
  REQUIRE(!failed);
}

TEST_CASE("One-dimensional tensor with wrong string count") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{tis};
  try {
    auto lt = T1("i", "i");
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
}

TEST_CASE("One-dimensional tensor with wrong label count") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{tis};
  try {
    auto lt = T1(tis.label(0), tis.label(1));
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
}

TEST_CASE("Two-dimensional tensor") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{tis, tis};
  try {
    auto lt = T1();
  } catch (...) {
    failed = true;
  }
  REQUIRE(!failed);
}

TEST_CASE("Two-dimensional tensor with correct string count") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{tis, tis};
  try {
    auto lt = T1("i", "j");
  } catch (...) {
    failed = true;
  }
  REQUIRE(!failed);
}

TEST_CASE("Two-dimensional tensor with correct label count") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{tis, tis};
  try {
    auto lt = T1(tis.label(0), tis.label(1));
  } catch (...) {
    failed = true;
  }
  REQUIRE(!failed);
}

TEST_CASE("Two-dimensional tensor with smaller string count") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{tis, tis};
  try {
    auto lt = T1("i");
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
}

TEST_CASE("Two-dimensional tensor with smaller label count") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{tis, tis};
  try {
    auto lt = T1(tis.label(0));
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
}

TEST_CASE("Two-dimensional tensor with larger string count") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{tis, tis};
  try {
    auto lt = T1("i", "j", "k");
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
}

TEST_CASE("Two-dimensional tensor with larger label count") {
  bool failed = false;
  IndexSpace is {range(10)};
  TiledIndexSpace tis{is};
  Tensor<double> T1{tis, tis};
  try {
    auto lt = T1(tis.label(0), tis.label(1), tis.label(2));
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
}

TEST_CASE("Two-dimensional dependent index tensor") {
  bool failed = false;
  IndexSpace is {range(1)};
  TiledIndexSpace tis{is};
  IndexSpace is2{range(5)};
  std::map<IndexVector, IndexSpace> dep_space_relation;
  dep_space_relation[{0}] = is2;
  IndexSpace dis{{tis}, dep_space_relation};
  TiledIndexSpace tdis{dis};

  TiledIndexLabel i, j, a, b;
  std::tie(i,j) = tis.labels<2>("all");
  std::tie(a,b) = tdis.labels<2>("all");

  //some tensor tests. @todo should be in Test_Tensor.cpp
  try {
    Tensor<double> T1{tdis, tdis};
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
  failed = false;

  try {
    Tensor<double> T1{a(i), j};
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
  failed = false;

  try {
    Tensor<double> T1{i(a), j};
  } catch (...) {
    failed = true;
  }
  REQUIRE(!failed);
  failed = false;


  //labeled tensor tests follow
  Tensor<double> T1({a(i), i});
  try {
    auto lt = T1();
  } catch (...) {
    failed = true;
  }
  REQUIRE(!failed);
  failed = false;

  try {
    auto lt = T1(b(j), j);
  } catch (...) {
    failed = true;
  }
  REQUIRE(!failed);
  failed = false;

  try {
    auto lt = T1("x");
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
  failed = false;

  try {
    auto lt = T1(a);
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
  failed = false;

  try {
    auto lt = T1(i);
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
  failed = false;

  try {
    auto lt = T1("x", "y");
  } catch (...) {
    failed = true;
  }
  REQUIRE(!failed);
  failed = false;

  try {
    auto lt = T1(a, i);
  } catch (...) {
    failed = true;
  }
  REQUIRE(!failed);
  failed = false;

  try {
    auto lt = T1(i, a(i));
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
  failed = false;

  try {
    auto lt = T1(a(i));
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
  failed = false;

  try {
    auto lt = T1(i);
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
  failed = false;

  try {
    auto lt = T1(i(a));
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
  failed = false;

  try {
    auto lt = T1(a(a));
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
  failed = false;

  try {
    auto lt = T1(i(i));
  } catch (...) {
    failed = true;
  }
  REQUIRE(failed);
  failed = false;
}

File: tests/python/Test_LabeledTensor.py
# test_zero_dim_tensor.py
#
# pytest translation of the doctest/C++ test.
#
# Note:
# The Python bindings wrap labeled tensors in PyLabeledTensor and do not always
# force C++ Tensor::operator() validation immediately for label-count cases.
# checked_lt() forces materialization through a harmless OpExpr construction so
# these tests match the C++ test intent.

import pytamm as tamm


def failed(fn):
    try:
        fn()
        return False
    except Exception:
        return True


def checked_lt(tensor, *args):
    lt = tensor(*args)

    # Force C++ labeled-tensor validation without executing anything.
    # This only constructs an OpExpr.
    tamm.copy(lt, lt)

    return lt


def make_tis(n=10):
    is_ = tamm.IndexSpace(tamm.range(n))
    return tamm.TiledIndexSpace(is_)


def tis_labels(tis, count):
    return tis.labels("all", None, count)


def test_zero_dimensional_tensor():
    T1 = tamm.TensorDouble()

    assert not failed(lambda: checked_lt(T1))


def test_zero_dimensional_tensor_wrong_str_count():
    T1 = tamm.TensorDouble()

    assert failed(lambda: checked_lt(T1, "a"))


def test_zero_dimensional_tensor_wrong_label_count():
    tis = make_tis(10)
    T1 = tamm.TensorDouble()

    # C++ uses T1(tis), relying on C++ conversion/overload behavior.
    # In Python bindings pass the explicit label.
    assert failed(lambda: checked_lt(T1, tis.label()))


def test_one_dimensional_tensor():
    tis = make_tis(10)
    T1 = tamm.TensorDouble([tis])

    assert not failed(lambda: checked_lt(T1))


def test_one_dimensional_tensor_with_correct_string_count():
    tis = make_tis(10)
    T1 = tamm.TensorDouble([tis])

    assert not failed(lambda: checked_lt(T1, "i"))


def test_one_dimensional_tensor_with_correct_label_count():
    tis = make_tis(10)
    T1 = tamm.TensorDouble([tis])

    assert not failed(lambda: checked_lt(T1, tis.label()))


def test_one_dimensional_tensor_with_wrong_string_count():
    tis = make_tis(10)
    T1 = tamm.TensorDouble([tis])

    assert failed(lambda: checked_lt(T1, "i", "i"))


def test_one_dimensional_tensor_with_wrong_label_count():
    tis = make_tis(10)
    T1 = tamm.TensorDouble([tis])

    assert failed(lambda: checked_lt(T1, tis.label(0), tis.label(1)))


def test_two_dimensional_tensor():
    tis = make_tis(10)
    T1 = tamm.TensorDouble([tis, tis])

    assert not failed(lambda: checked_lt(T1))


def test_two_dimensional_tensor_with_correct_string_count():
    tis = make_tis(10)
    T1 = tamm.TensorDouble([tis, tis])

    assert not failed(lambda: checked_lt(T1, "i", "j"))


def test_two_dimensional_tensor_with_correct_label_count():
    tis = make_tis(10)
    T1 = tamm.TensorDouble([tis, tis])

    assert not failed(lambda: checked_lt(T1, tis.label(0), tis.label(1)))


def test_two_dimensional_tensor_with_smaller_string_count():
    tis = make_tis(10)
    T1 = tamm.TensorDouble([tis, tis])

    assert failed(lambda: checked_lt(T1, "i"))


def test_two_dimensional_tensor_with_smaller_label_count():
    tis = make_tis(10)
    T1 = tamm.TensorDouble([tis, tis])

    assert failed(lambda: checked_lt(T1, tis.label(0)))


def test_two_dimensional_tensor_with_larger_string_count():
    tis = make_tis(10)
    T1 = tamm.TensorDouble([tis, tis])

    assert failed(lambda: checked_lt(T1, "i", "j", "k"))


def test_two_dimensional_tensor_with_larger_label_count():
    tis = make_tis(10)
    T1 = tamm.TensorDouble([tis, tis])

    assert failed(lambda: checked_lt(T1, tis.label(0), tis.label(1), tis.label(2)))


def test_two_dimensional_dependent_index_tensor():
    is_ = tamm.IndexSpace(tamm.range(1))
    tis = tamm.TiledIndexSpace(is_)

    is2 = tamm.IndexSpace(tamm.range(5))

    dep_space_relation = {
        (0,): is2,
    }

    dis = tamm.IndexSpace([tis], dep_space_relation)
    tdis = tamm.TiledIndexSpace(dis)

    i, j = tis_labels(tis, 2)
    a, b = tis_labels(tdis, 2)

    # Tensor construction tests.
    assert failed(lambda: tamm.TensorDouble([tdis, tdis]))

    assert failed(lambda: tamm.TensorDouble([a(i), j]))

    assert not failed(lambda: tamm.TensorDouble([i(a), j]))

    # Labeled tensor tests.
    T1 = tamm.TensorDouble([a(i), i])

    assert not failed(lambda: checked_lt(T1))

    assert not failed(lambda: checked_lt(T1, b(j), j))

    assert failed(lambda: checked_lt(T1, "x"))

    assert failed(lambda: checked_lt(T1, a))

    assert failed(lambda: checked_lt(T1, i))

    assert not failed(lambda: checked_lt(T1, "x", "y"))

    assert not failed(lambda: checked_lt(T1, a, i))

    assert failed(lambda: checked_lt(T1, i, a(i)))

    assert failed(lambda: checked_lt(T1, a(i)))

    assert failed(lambda: checked_lt(T1, i))

    assert failed(lambda: checked_lt(T1, i(a)))

    assert failed(lambda: checked_lt(T1, a(a)))

    assert failed(lambda: checked_lt(T1, i(i)))